File: Python/dynamic_programming/minimum_coins.py
import sys


def min_coins(coins, target):
    dp = [sys.maxsize] * (target + 1)
    dp[0] = 0
    for i in range(1, target + 1):
        for coin in coins:
            if coin <= i:
                rem_target = 1 + dp[i - coin]
                if rem_target != sys.maxsize:
                    dp[i] = min(dp[i], rem_target)
    return dp[target]


if __name__ == "__main__":
    print(min_coins([10, 5, 25], 30))

File: C++/Recursion/MinimumCoins.cpp
#include <bits/stdc++.h>
using namespace std;

int minCoins(int coins[], int n, int target)
{
  if (target == 0)
    return 0;
  int res = INT32_MAX;
  for (int i = 0; i < n; i++)
  {
    if (coins[i] <= target)
    {
      int remTarget = 1 + minCoins(coins, n, target - coins[i]);
      if (remTarget != INT32_MAX)
        res = min(res, remTarget);
    }
  }
  return res;
}

int main()
{
  int coins[] = {10, 25, 5};
  int n = sizeof(coins) / sizeof(coins[0]);
  cout << minCoins(coins, n, 30) << endl;
}